### PR TITLE
Protocol change - Players start with each others' UTXOs, fee changes

### PR DESCRIPTION
**Pull request by Crypt-iQ**
Each player knows every other player's UTXOs before the shuffle starts.  Also, each player can have a different fee based on the the number of UTXOs they choose to shuffle.

### CODE REVIEW

File: shuffler/src/main/java/com/shuffle/bitcoin/Signatures.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.shuffle.bitcoin;
+
+import com.shuffle.p2p.Bytestring;
+
+import java.util.HashSet;
+
+/**
+ * Created by nsa on 11/13/16.
+ */
+public class Signatures extends Bytestring {
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/bitcoin/Signatures.java]: Don't extend Bytestring. A Bytestring is just one bytestring. A Signatures is a set of Bytestrings. Therefore a Signatures is not a Bytestring.

File: shuffler/src/main/java/com/shuffle/bitcoin/Transaction.java [hunk 2]
```diff
@@ -12,6 +12,7 @@
 import com.shuffle.protocol.FormatException;
 
 import java.io.Serializable;
+import java.util.HashSet;
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/bitcoin/Transaction.java]: You don't need to add this import.

**Comment by DanielKrawisz** [comment]
Can you not add BitcoinCore.java in this pr? That's not logically related to "Protocol change - players exchange UTXO lists rather than addresses" Also please provide a detailed description of this pr.

File: shuffler/src/main/java/com/shuffle/sim/Adversary.java [hunk 3]
```diff
@@ -97,9 +97,12 @@ public void run() {
 
                 try {
                     try {
+                        /*
                         q.send(new Either<>(shuffle.runProtocol(
                                 amount, fee, sk, players, anon, change, null
-                        ), null));
+                        ), null));*/
+                        // TODO
+                        q.send(new Either<>(shuffle.runProtocol(0, 0, null, null, null, null, null, null), null));
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/sim/Adversary.java]: Well this change will make the protocol not work.

File: shuffler/src/main/java/com/shuffle/sim/InitialState.java [hunk 4]
```diff
@@ -276,9 +276,13 @@ public Adversary adversary(
                         )
                 );
             } else if (mutate) {
-                shuffle = new CoinShuffle(messages, testCase.crypto(), coin.mutated());
+                // TODO
+                shuffle = null;
+                //shuffle = new CoinShuffle(messages, testCase.crypto(), coin.mutated());
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/sim/InitialState.java]: Here are some more changes that obviously would make the protocol not work.

File: shuffler/src/main/java/com/shuffle/sim/InitialState.java [hunk 5]
```diff
@@ -276,9 +276,13 @@ public Adversary adversary(
...
             } else {
-                shuffle = new CoinShuffle(messages, testCase.crypto(), coin);
+                // TODO
+                shuffle = null;
+                //shuffle = new CoinShuffle(messages, testCase.crypto(), coin);
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/sim/InitialState.java]: And again.

File: shuffler/src/test/java/com/shuffle/bitcoin/blockchain/BlockCypherDotComTest.java [hunk 6]
```diff
@@ -13,6 +13,8 @@
  * Created by conta on 25.07.16.
  */
 public class BlockCypherDotComTest {
+   // TODO
+   /*
```
Comment (DanielKrawisz) [inline shuffler/src/test/java/com/shuffle/bitcoin/blockchain/BlockCypherDotComTest.java]: Don't comment out tests. The tests should all work.

File: shuffler/src/test/java/com/shuffle/bitcoin/blockchain/TestBitcoin.java [hunk 7]
```diff
@@ -26,7 +26,10 @@
 /**
  * Created by Daniel Krawisz on 7/26/16.
  */
+
 public class TestBitcoin {
+    // TODO
+    /*
```
Comment (DanielKrawisz) [inline shuffler/src/test/java/com/shuffle/bitcoin/blockchain/TestBitcoin.java]: don't comment out tests.

File: shuffler/src/test/java/com/shuffle/bitcoin/blockchain/TestBtcd.java [hunk 8]
```diff
@@ -9,36 +9,30 @@
...
 
+        boolean isUTXO = test.isUtxo("fe68ebbbec06746a26d4038b500b3b9450b91f9e5c0935304b18fe13ec3eccbd", 10);
+        System.out.println(isUTXO);
+
+    }
+
+    // TODO
+    /*
+
+    NetworkParameters netParams = MainNetParams.get();
```
Comment (DanielKrawisz) [inline shuffler/src/test/java/com/shuffle/bitcoin/blockchain/TestBtcd.java]: Don't comment out tests.

**Comment by DanielKrawisz** [comment]
There are a ton of files where the only changes are removing tabs and the newlines at the ends of files. Please figure out what is causing those changes and turn it off and get rid of all of the changes. They make it really difficult to read through this pr.

File: shuffler/src/main/java/com/shuffle/player/proto/Proto.java [hunk 9]
```diff
@@ -6,13 +6,13 @@
 public final class Proto {
   private Proto() {}
   public static void registerAllExtensions(
-      com.google.protobuf.ExtensionRegistry registry) {
+          com.google.protobuf.ExtensionRegistry registry) {
```
Comment (jonathancross) [inline shuffler/src/main/java/com/shuffle/player/proto/Proto.java]: > Daniel K shouldn't look at this file at all.

Makes sense.

File: shuffler/src/main/java/com/shuffle/sim/Adversary.java [hunk 10]
```diff
@@ -97,12 +97,12 @@ public void run() {
...
                     try {
-                        q.send(new Either<>(shuffle.runProtocol(
-                                amount, fee, sk, players, anon, change, null
+                        q.send(new Either<Transaction, Matrix>(shuffle.runProtocol(
+                                amount, fee, sk, players, null, anon, change, null
                         ), null));
 
                     } catch (Matrix m) {
-                        q.send(new Either<>(null, m));
+                        q.send(new Either<Transaction, Matrix>(null, m));
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/sim/Adversary.java]: You didn't need to add this in.

File: shuffler/src/test/java/com/shuffle/bitcoin/blockchain/TestBitcoin.java [hunk 11]
```diff
@@ -122,7 +129,8 @@ public MockBitcoin() {
     // TODO
     @Test
     public void testSufficientFunds() throws Exception {
-        MockBitcoin mock = new MockBitcoin();
+        //MockBitcoin mock = new MockBitcoin();
+        Btcd mock = new Btcd(TestNet3Params.get(), "admin", "pass");
```
Comment (DanielKrawisz) [inline shuffler/src/test/java/com/shuffle/bitcoin/blockchain/TestBitcoin.java]: This variable should no longer be called mock since it's no longer a mock object.

File: shuffler/src/main/java/com/shuffle/player/proto/message.proto [hunk 12]
```diff
@@ -35,9 +35,10 @@ message Message {
     EncryptionKey key = 2;
     Hash hash = 3;
     Signature signature = 4;
-    string str = 5;
-    Blame blame = 6;
-    Message next = 7;
+    Signatures signatures = 5;
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/player/proto/message.proto]: Woah woah woah! This isn't how you're supposed to use protobuf! You should always add a NEW number to the end if you're adding a new field. Signatures should be number 8.

File: shuffler/src/main/java/com/shuffle/player/Player.java [hunk 13]
```diff
@@ -171,7 +177,7 @@ public Running start() throws IOException, InterruptedException {
                     | AddressFormatException
                     | ExecutionException e) {
 
-                stream.println("  Player " + sk.VerificationKey() + " reports error " +  e.getMessage());
+                stream.println("   Player " + sk.VerificationKey() + " reports error " +  e.getMessage());
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/player/Player.java]: Why did you add an extra space here?

File: shuffler/src/main/java/com/shuffle/player/Player.java [hunk 14]
```diff
@@ -72,6 +74,8 @@
 
     private final SortedSet<VerificationKey> addrs;
 
+    private final Map<VerificationKey, HashSet<TransactionOutPoint>> outs;
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/player/Player.java]: This variable name is very confusing. Yes, this is a list of outputs, but it's not clear at all how they are to be used in the protocol. How about naming it 'funded' or 'redeemable' ? Whatever name you choose change it throughout the whole pr.

File: shuffler/src/main/java/com/shuffle/player/Message.java [hunk 15]
```diff
@@ -177,6 +192,8 @@ public boolean equals(Object o) {
 
             Atom a = (Atom) o;
 
+            // TODO
+            // a.sigs == a.sigs
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/player/Message.java]: It looks like you should uncomment these lines and similarly commented lines later on.

File: shuffler/src/main/java/com/shuffle/player/Message.java [hunk 16]
```diff
@@ -69,6 +72,7 @@ public Atom(
                 EncryptionKey ek,
                 SecureHash secureHash,
                 Bytestring sig,
+                Signatures sigs,
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/player/Message.java]: We're not using Bytestring sig at all anymore, right? If not you should remove it.

File: shuffler/src/main/java/com/shuffle/p2p/Connect.java [hunk 17]
```diff
@@ -151,6 +151,8 @@ public int increment(Identity identity) {
     public Connect(Channel<Identity, P> channel, Crypto crypto)
             throws InterruptedException, IOException {
 
+        // TODO
+        // For large Shuffles, a capacity of 100 could be too low
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/p2p/Connect.java]: Good point. Let's remember to add a constant called MAX_PLAYERS and set it at 100 and then use that instead of using this magic number here! The program should fail gracefully if you try to enter too many.

File: shuffler/src/main/java/com/shuffle/mock/TransactionMutator.java [hunk 18]
```diff
@@ -35,15 +35,17 @@ public TransactionMutator(MockCoin coin) {
     public Transaction shuffleTransaction(
             long amount,
             long fee,
+            Map<VerificationKey, Long> feeEstimates,
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/mock/TransactionMutator.java]: These shouldn't be estimates. Everyone has to know exactly what the fee is going to be at the outset.

File: shuffler/src/main/java/com/shuffle/player/Shuffle.java [hunk 19]
```diff
@@ -83,115 +86,11 @@
...
     private static long MIN_AMOUNT = 1000000;
-    // 5000 satoshis
-    private static long MIN_FEE = 5000L;
+    // 15 satoshis per byte
+    private static long MIN_FEE = 15L;
 
     // Entropy checker must think there is at least this much entropy.
     private static int MIN_APPARENT_ENTROPY = 128;
-
-    public static OptionParser getShuffleOptionsParser() {
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/player/Shuffle.java]: Put the functions in this file back in the order they were. Otherwise it's really hard to tell what changed in this file!

File: shuffler/src/main/java/com/shuffle/protocol/CoinShuffle.java [hunk 20]
```diff
@@ -115,14 +124,20 @@ Transaction protocolDefinition(
...
             phase.set(Phase.Announcement);
             log.info("Player " + me + " begins CoinShuffle protocol " + " with " + N + " players.");
-            System.out.println("Player " + me + " begins CoinShuffle protocol " + " with " + N + " players.");
 
             // Check for sufficient funds.
             // There was a problem with the wording of the original paper which would have meant
             // that player 1's funds never would have been checked, but it's necessary to check
             // everybody.
+
+            // estimate fee per player
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/protocol/CoinShuffle.java]: We shouldn't be estimating fees here. The fees need to be already known at the time the program starts running.

File: shuffler/src/main/java/com/shuffle/protocol/CoinShuffle.java [hunk 21]
```diff
@@ -132,8 +147,8 @@ Transaction protocolDefinition(
             // needs to send an announcement message at this point too because he might
             // have a change address. Therefore he just follows the same procedure as
             // everyone else.
+            // Additionally, each player broadcasts their utxo list.
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/protocol/CoinShuffle.java]: The list of utxos for all players should be known before the program starts running, so they shouldn't do that.

File: shuffler/src/main/java/com/shuffle/protocol/CoinShuffle.java [hunk 22]
```diff
@@ -477,9 +496,18 @@ private void blameInsufficientFunds()
 
             // Check that each participant has the required amounts.
             for (VerificationKey player : players.values()) {
-                if (!coin.sufficientFunds(player.address(), amount + fee)) {
-                    // Enter the blame phase.
-                    offenders.add(player);
+                if (feeEstimates != null) {
+                    Address a = new AddressUtxoImpl(fundedOutputs.get(player));
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/protocol/CoinShuffle.java]: This logic should all take place in sufficientFunds, not in the main protocol function.

File: shuffler/src/main/java/com/shuffle/protocol/CoinShuffle.java [hunk 23]
```diff
@@ -519,12 +547,18 @@ void blameBroadcastShuffleMessages()
...
             Message doubleSpend = messages.make();
             for (VerificationKey key : players.values()) {
-                Transaction o = coin.getConflictingTransaction(t, key.address(), amount);
+                Address a;
+                if (fundedOutputs != null) {
+                    a = new AddressUtxoImpl(fundedOutputs.get(key));
+                } else {
+                    a = key.address();
+                }
+                Transaction o = coin.getConflictingTransaction(t, a, amount);
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/protocol/CoinShuffle.java]: Same thing here. All this logic should take place in getConflictingTransaction.

**Comment by DanielKrawisz** [comment]
feeEstimates should not be an input to the runProtocol function. We should just already know what each player has to pay beforehand. Instead, how about an input which is a map from the VerificationKey to a specific amount that each player has to pay?

**Comment by Crypt-iQ** [comment]
Cleaning up this branch.
